RFD: WebAuthn Support

codingllama opened the pull request:
RFD for WebAuthn support on Teleport.

codingllama committed:
Initial draft for the WebAuthn RFD

codingllama commented:
Preview link: https://github.com/gravitational/teleport/blob/codingllama/rfd-webauthn/rfd/9999-webauthn-support.md.

codingllama commented:
Note for readers: the current RFD covers mostly the server-side parts of WebAuthn support, as I still have some unknowns left to investigate for client-side (especially concerning `tsh`). That said, I think it's likely most of the server-side design will stand, so please let me know your thoughts on that.

FYI @awly @russjones @klizhentas

awly commented on rfd/9999-webauthn-support.md:
````diff
@@ -0,0 +1,501 @@
+```
+
+The key space is `web/users/{user_id}/webauthnsessiondata/{session_id}`.
+Implementation may choose to use hard-coded session IDs for login and
+registration, limiting the possibility of concurrent attempts, or pick distinct
+IDs, as appropriate.
````
u2f uses the `user_id` without `session_id` to store the challenge
it seems to work well enough despite limiting concurrent logins

codingllama replied:
Ack.

I'll keep this one - code complexity is pretty much the same, it's not difficult to read and it allows us to fine-tune behavior better.

awly reviewed (commented):
Meta-comment on configuration: make it very easy to enable webauthn for admins who know nothing about the protocol or the differences between u2f and webauthn.
Ideally, admin already uses `second_factor: on`, upgrades to teleport v7.x (with this feature implemented) and webauthn devices "just work".
If it's impossible, they should have to specify the minimum number of fields to make it work.

All the situational webuathn options we should choose ourselves internally and not make the cluster admin decide.

codingllama commented:
> Meta-comment on configuration: make it very easy to enable webauthn for admins who know nothing about the protocol or the differences between u2f and webauthn.
Ideally, admin already uses second_factor: on, upgrades to teleport v7.x (with this feature implemented) and webauthn devices "just work".

Agreed. In the current design, if they have second_factor: on and upgrade, the backend will start sending WebAuthn challenges, which the frontends should favor and it all silently works. In browsers we'll get WebAuthn/CTAP2 devices for free, in the CLI we'll have to put some work (but it should be transparent too).

> All the situational webuathn options we should choose ourselves internally and not make the cluster admin decide.

Sounds good - I'll keep a note to update the design with simplified configurations and call out any important defaults (from the top of my head, attestation mode and user presence are the main ones).

codingllama commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,501 @@
+solution for this problem and seems to be well-supported for browsers
+(__TODO(codingllama):__ Verify for Firefox and Safari).
+
+Unfortunately, the appid extension is not without its issues.
+[As the standard itself says in item 2](https://www.w3.org/TR/webauthn/#sctn-appid-extension),
+_"When verifying the assertion, expect that the rpIdHash MAY be the hash of the
+AppID instead of the RP ID."_ This advice
+[wasn't followed by github.com/duo-labs/webauthn](https://github.com/duo-labs/webauthn/blob/9f1b88ef44cc0e4f5ddf511ed12a3aa468f972d7/protocol/assertion.go#L117),
+but in the short-term may be circumvented by creating a secondary
+[WebAuthn object](https://pkg.go.dev/github.com/duo-labs/webauthn/webauthn#WebAuthn)
+with RFID = AppID, used exclusively to validate legacy requests (identified by
+the presence of the "incorrect" rpIdHash).
```
WebAuthn has distinct fields for the RPID and the Origin, so, under normal circumstances, only the cluster that emitted the challenge should match the origin (host+port).

Something I also considered, but failed to mention, is that we can tell whether the device used to perform MFA is a pre-WebAuthn device - only in this case we should enable the AppID fallback.

We are able to cover the essentials here: user presence and attestation.

codingllama replied:
Let me know if you'd like me to capture some of that reply into the RFD text.

codingllama commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,501 @@
+    second_factor: webauthn # "on" allows all MFA options
+    webauthn:
+      relying_party:
+        display_name: "Teleport Auth"
```
I suspect some authenticators may store the display_name along with Resident Keys, but I haven't seen it being used yet.

codingllama replied:
Note to self: we may be able to get away without those settings, or with only hardcoded values for them.

codingllama commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,501 @@
+      # previously-registered U2F devices, to be used in place of the Relying
+      # Party ID, thus keeping backwards compatibility.
+      # Not required for new MFA installations.
+      # Defaults to "auth_service.authentication.u2f.app_id", if present.
+      u2f_app_id: "https://localhost:3080"
```
I explicitly included this one because it allows users moving from u2f to webauthn to delete the entire u2f section of the config, keeping everything related to webauthn here.

klizhentas replied:
@codingllama what is the resource version of it? (we support two modes - static config and the dynamic resource for cloud and K8s and terraform). Is there a way for us to avoid extra field?

codingllama replied:
The resource version for `AuthPreferenceSpecV2` doesn't have to change, we are only adding fields to it. Is that OK or do we have a different policy?  (The proto file leads me to believe we are at v2).

We could avoid the u2f_app_id field (more details about that in the "second_factor: u2f" thread). Taken to an extreme, we could avoid the entire webauthn configuration stanza if we are ok with keeping u2f.app_id indefinitely (and removing a few optional customization options, like the RP display_name and icon). As it stands it is optional anyway. I'd rather provide the toggles, but I won't lose any sleep if they're gone either  :)

codingllama commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,501 @@
+      # github.com/duo-labs/webauthn library, which may be narrower than the
+      # full WebAuthn spec.
+
+      # Spec:
+      # https://www.w3.org/TR/webauthn/#dictdef-authenticatorselectioncriteria
+      authenticator_selection:
+        authenticator_attachment: ""                # "platform" or "cross-platform", empty means "no attachment"
+        require_resident_key: no                    # "yes" or "no"
+        user_verification_requirement: "preferred"  # "required", "preferred" or "discouraged"
+      timeout_ms: "60000"  # Default timeout for WebAuthn operations, in milliseconds
```
Sounds good. Defaults are likely to be the values I had exemplified there: authenticators may be "platform"(touch ID) or "cross-platform" (yubikeys), resident keys not required, user verification preferred (but not required).

codingllama commented on rfd/9999-webauthn-support.md:
````diff
@@ -0,0 +1,501 @@
+      silence_u2f_compatibility_warning: no  # see below.
+```
+
+It's important to note that, while WebAuthn features are well-supported by web
+browsers, CLI implementations may be unable to conform to certain WebAuthn
+settings, in particular when using FIDO/CTAP1 libraries. For example, while
+`tsh` relies on [github.com/flynn/u2f](https://github.com/flynn/u2f), we may be
+unable to enforce certain authenticator_selection settings.
+
+Furthermore, when migrating from U2F to WebAuthn, the creation of a more strict
+policy may prohibit the use of previously-registered devices (for example,
+"authenticator_attachment: platform" may prohibit previously-registered
+YubiKeys from working). In cases such as this the Auth server will issue a
+warning on startup, but ultimately the configuration stands.
+
+
+__TODO(codingllama):__ Support attestation.
````
Agreed.

codingllama commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,501 @@
+[wasn't followed by github.com/duo-labs/webauthn](https://github.com/duo-labs/webauthn/blob/9f1b88ef44cc0e4f5ddf511ed12a3aa468f972d7/protocol/assertion.go#L117),
+but in the short-term may be circumvented by creating a secondary
+[WebAuthn object](https://pkg.go.dev/github.com/duo-labs/webauthn/webauthn#WebAuthn)
+with RFID = AppID, used exclusively to validate legacy requests (identified by
```
Done. Well spotted.

codingllama committed:
Address review comments: wording and spelling

codingllama committed:
Simplify configuration settings

codingllama commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,501 @@
+        # Defaults to the Proxy domain name.
+        # IMPORTANT: similarly to the U2F app ID, if the Relying Party ID
+        # changes then all devices must be re-registered.
+        id: localhost
```
Considering how both CLI and Proxy UI talk to the Proxy for login, it's difficult to see this being different (technically you could use the Auth server both for ID and Origin, but that would fail all our current tools/interfaces).

It's important for WebAuthn, especially when browser-based, that both RPID and Origin match the backend returning/verifying the challenges. The implications of this setting are that authn goes through the Proxy, which sounds correct to me.

Simplified configuration to omit RPID.

codingllama commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,501 @@
+        # Origin of the Relying Party, validated against the origin of
+        # authentication requests.
+        # Defaults to the Proxy public address.
+        origin: https://localhost:3080
```
Done, see earlier comments for explanation.

codingllama commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,501 @@
+1. Initially, `tsh` keeps support for only FIDO/CTAP1 (via github.com/flynn/u2f),
+   but is changed to use WebAuthn-based endpoints
+
+2. At a later stage (to be designed), support is brought in for FIDO2 / Touch
+   ID.
```
Will do as soon as it's clearer, but it likely depends on whatever libraries we choose to adopt.

The abridged answer I currently have is
* USB-based CTAP1 backends we currently support while running [flynn/u2f](https://github.com/flynn/u2f#compatibility) (this is a given)
* USB-based CTAP1/2 backends [libfido2](https://github.com/Yubico/libfido2) supports, if/when we go that route
* Apple biometrics/Touch ID support is likely to be hand-crafted, supporting all models that have it (APIs seem to be essentially the same)
* Windows biometrics TBD (likely hand-crafted if I were to hazard a guess)
* NFC and other transports TBD

codingllama committed:
Make tsh rely on the /webapi/ping server version

codingllama commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,501 @@
+keeping the old endpoints in place. Additions to existing messages are unknown
+by legacy code, thus ignored by older Proxy and `tsh` versions.
+
+The `/webapi/ping` response may be modified to signalize to `tsh` that the Proxy
+supports the new /mfa/ endpoints (with the assumption, by the Proxy, that the
+Auth server also supports the new endpoints). This allows newer `tsh` versions
+to safely talk to older Proxies (at least for a few releases).
```
Done.

I had considered using the server version, decided against it so we didn't need to bake in specific versions into code (less readable than a custom setting), but on second thought it seems better to have a simpler ping reply.

codingllama committed:
Add a section on user handles

codingllama committed:
Add the Attestation section

codingllama committed:
Add a references section

codingllama commented:
PTAL - addressed comments and added sections on user handles and attestation.

codingllama committed:
Update tsh+WebAuthn with latest data

codingllama committed:
Add detailed Touch ID sections

codingllama commented:
Hey everyone, I just added a few sections related to Touch ID (ctrl+f `# Touch ID`), please let me know your thoughts. (@klizhentas, I think this may interest you.)

As I was saying to Russell earlier today, I think we've done enough prep work, so I'll move ahead with the server-side implementation (which I think is not controversial, plus we need it to have everything going). I am happy to retract/change/delete whatever I push depending on the comments we get here.

FYI @awly @russjones

codingllama committed:
Errata on App Attestation Service

It appears to only support iOS.

klizhentas commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,674 @@
+previously-registered devices and keep their workflow unchanged.
+
+Migration from U2F to WebAuthn is achieved simply by changing the second_factor
+option from "u2f" to "webauthn". The WebAuthn backend sends challenges for all
```
@codingllama if webauthn is a superset of `u2f`, why not make update automatic, making u2f mean the same as webauthn and removing the u2f from the documentation in the future?

codingllama replied:
The main issue I see are Teleport version rollbacks and their corner-cases. If you upgrade to a version where "u2f" means "webauthn" and later rollback, the devices registered in-between may not be true U2F devices (they may disappear, not work, etc, depending on how we go about this). I am concerned this might break user expectation if they never touched their configs during the update (vs someone who actually changed from "u2f" to "webauthn").

Let's say we do this, then the question is what are the acceptable "breakages" in case of a rollback.

* FIDO2 registered devices would "disappear" from storage (as they are saved as WebAuthnDevices)
* U2F devices registered after the upgrade would either disappear (if saved as WebAuthnDevice) or would require special system handling to keep working (ie, saving as both WebAuthnDevice and U2FDevice)

Newer devices (eg, yubikey 5+) are all likely to disappear, even if they do support U2F, as I assume they would favor the FIDO2 protocol over U2F/CTAP1.

One could also argue that this is due to happen anyway when u2f is removed/deprecated, in which case we might do it now. Devices disappearing on rollbacks would also be the current behavior for "second_factor: on".

WDYT?

klizhentas replied:
@codingllama sounds good, for the transition period let's say for 8.0 we can keep `webauthn` and `u2f` modes separate, but will replace all mentions in the documentation with webauthn going forward. Then in the next major verison we should make them aliases. If you agree, let's note this in the RFD

klizhentas commented on rfd/9999-webauthn-support.md:
why not switch this to GRPC + tls SNI? this will let the client talk to the auth server and not the proxy

codingllama replied:
I thought you might suggest this. :)

I agree that switching to gRPC is desirable, but I am concerned about changing too much at once. I'd rather push webauthn support first, then refactor the login in a separate change. Moving it all to gRPC is a larger refactor that will touch a good chunk of the local login flow, whereas webauthn support is a more focused change with little impact to existing endpoints.

klizhentas replied:
Sounds good to me, let's just create a ticket, so we don't forget.

codingllama replied:
Done - https://github.com/gravitational/teleport/issues/7966.

klizhentas reviewed (commented):
This is solid, super thorough overview and design research 👍 

I'm assuming this is still a draft, but here are a couple of early comments:

* Let's find a way to reduce amount of parameters, e.g make u2f a config alias for webauthn, so we can deprecate u2f in the future.
* Evaluate scenarios when configuration will be done by resources using tctl rm/create , see https://github.com/gravitational/teleport/blob/master/rfd/0028-cluster-config-resources.md for details
* Why not take advantage of @smallinsky switch to TLS SNI and switch to the client talking over GRPC directly via the auth server, instead of duplicating POST calls via proxy? This should be available soon. The client can dial the auth server via proxy by using TLS ALPN Next proto. 
* Can you clarify the part about app id differences and new fields from the server administrator's perspective?

codingllama commented:
Thanks for the review, @klizhentas. I'll update the PR from draft to open, as I don't have to plans to change the design further (unless due to comment feedback).

Upgrading to webauthn, in terms of configuration, works like this:

* If they have "second_factor: on", then WebAuthn takes over from U2F, using the app ID from the u2f config as necessary
* If they have "second_factor: u2f", they would have to change it to "second_factor: webauthn". The app ID is taken from the u2f config, as above

In regards to the U2F app ID, we can't really guess it. I included the field in the webauthn configuration so users may completely remove the u2f configuration from the file if they so desire. We could take this  field out of the design and always refer to the u2f config for the app ID, but this means that that part of the config will outlive even the eventual removal of the u2f code.

codingllama committed:
Simplify config and expand on RPID and RPOrigin

codingllama commented:
Hey everyone,

I've pushed a few config simplifications in response to your comments. A quick sum-up of changes is:

* Removed RPDisplayName (defaults to "Teleport") and RPIcon (defaults to "", could change to a webassets icon too).
* Removed RPOrigin - we assert it based on the RPID (initial impl) and known Proxy/Auth addresses (improved impl)
* RPID is optional for U2F-enabled configs (we derive it from u2f.app_id), mandatory for new installs

I had another almost-complete change that went around detailing how to infer the RPID, how we would store the inferred value (so it doesn't suddenly change after used), how to deal with ambiguities (or fail in face of them) and so on. It was getting a bit complex and had bothersome corner-cases. Then it dawned of me that the U2F app_id is a mandatory parameter for most second_factor modes, so now we take advantage of that fact to make upgrades to webauthn simple to do (and to code!)

Looking forward to your feedback.

@klizhentas @awly @russjones

klizhentas commented on rfd/9999-webauthn-support.md:
```diff
@@ -0,0 +1,708 @@
+WebAuthn challenges. In this case user-facing interfaces, such as `tsh` and the
+Proxy UI, should favor replying to WebAuthn challenges (instead of U2F).
+
+Additions to `teleport.yaml` are as follows:
```
Note: need a resource `cluster_auth_preference` example conterpart to this static configuration, as this won't work for cloud setup. See cloud docs here:

https://goteleport.com/docs/cloud/guides/u2f/

codingllama replied:
Noted, thanks for the reference.

codingllama committed:
Only use attestation from U2F if all attestation keys are empty

codingllama committed:
Add U2F phase out section

codingllama committed:
Tweaks to names and messages to better match the codebase

codingllama committed:
Mention cluster_auth_preference changes

klizhentas committed:
Merge branch 'master' into codingllama/rfd-webauthn
